### PR TITLE
[SYCL][CUDA] Implemented cuda_piextUSMEnqueueMemAdvise

### PR DESCRIPTION
**Pull request by nyalloc**
This PR implements the USM MemAdvise function for CUDA backend.

The  `PI_COMMAND_TYPE_MEMADVISE` enumerator is added, the value of which is derived from `CL_COMMAND_MEMADVISE_INTEL` constant as defined in the preview [USM extension specification](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc).

### CODE REVIEW

**Comment by nyalloc** [comment]
Hello @smaslov-intel ! Hope you are well. This PR seems to have flown under the radar - can we do anything to move this along?

Cheers.

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -390,7 +390,8 @@ typedef enum {
   PI_COMMAND_TYPE_SVM_MEMCPY = CL_COMMAND_SVM_MEMCPY,
   PI_COMMAND_TYPE_SVM_MEMFILL = CL_COMMAND_SVM_MEMFILL,
   PI_COMMAND_TYPE_SVM_MAP = CL_COMMAND_SVM_MAP,
-  PI_COMMAND_TYPE_SVM_UNMAP = CL_COMMAND_SVM_UNMAP
+  PI_COMMAND_TYPE_SVM_UNMAP = CL_COMMAND_SVM_UNMAP,
+  PI_COMMAND_TYPE_MEMADVISE = 0x4207
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:394]: Do you absolutely need to add new true command type? Can't just use "fake" `PI_COMMAND_TYPE_USER`?

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:394]: If we are to extend the enumerations then we need to do it in a more robust (guarantee no overlap) and consistent (different backends need to handle those) way.

Comment (nyalloc) [reply sycl/include/CL/sycl/detail/pi.h:394]: I could do this yes, but I felt it was a good idea to do this as it is a defined value in intel's USM extension document and re-usable by the OpenCL backend. But if you'd rather we move away from that then that's totally fine with me, happy to replace the usage with `PI_COMMAND_TYPE_USER`.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:394]: Let's stick to `PI_COMMAND_TYPE_USER` until we really need to distinguish. FWIW, this is also what Level-Zero plugin is using: https://github.com/intel/llvm/blob/0c0f4c5cd729fc9060fa4c7fd3ce5646cdf94f3e/sycl/plugins/level_zero/pi_level_zero.cpp#L5804

**Review by smaslov-intel — COMMENTED** [review]
Please address comment

**Comment by nyalloc** [comment]
> Please address comment

No problem, happy to fix this. I'll update the PR today.

File: sycl/include/CL/sycl/detail/pi.h (hunk 1 above)
Comment (nyalloc) [reply sycl/include/CL/sycl/detail/pi.h]: Done

**Comment by nyalloc** [comment]
@smaslov-intel Updated. Thanks for the feedback.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -4449,9 +4449,30 @@ pi_result cuda_piextUSMEnqueueMemAdvise(pi_queue queue, const void *ptr,
...
+    result = PI_CHECK_ERROR(
+        cuMemAdvise((CUdeviceptr)ptr, length, (CUmem_advise)advice,
+                    queue->get_context()->get_device()->get()));
+    if (event) {
+      result = event_ptr->record();
+      *event = event_ptr.release();
+    }
+  } catch (pi_result err) {
+    result = err;
+  }
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/pi_cuda.cpp:4474]: please `catch(...)` too such that plugin never throws

Comment (nyalloc) [reply sycl/plugins/cuda/pi_cuda.cpp:4474]: Can do.

**Review by smaslov-intel — APPROVED** [review]
Thanks!

**Pull request merged**